rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,229 @@
+1. Flags will be primitive and will not allow complex data.
+1. Nested flags and flags with metadata will need hacks.
+
+##### Considerations
+
+Feature flag env vars can be prefixed with a string like `FF_` to separate them from the other env vars.
+
+The limitation for the data to be a string means any extra metadata must be JSON-stringified and then parsed in the code. Only primitive flag states can be handled conveniently.
+
+Nesting flags will use hacks like artificial separators e.g. `FF_PARENT__CHILD`. Looks ugly and confusing.
+
+#### 2. Query params
+
+##### Pros
+
+1. Even more effortless to set up.
+1. We've done this before for the embedded mode, [_remember_](https://github.com/WordPress/openverse-frontend/pull/65)?
+1. Testing is easy because tests can append the right flag param to the URL.
+
+##### Cons
+
+1. Configuration will need to live in the codebase so change of state will need deployment.
+1. Exposes half-baked features to users who know the flag.
+1. Transient, so flags must be set on every URL or stored in cookies.
+1. Flags will be primitive and will not allow complex data. Even stringified JSON would be bad in URLs.
+
+##### Considerations
+
+We can clearly demarcate feature flag params with a prefix such as `ff_` (or `feature/` 🚳) to separate them from the actual query params and process them separately.
+
+Due to their nature being run-time rather than build-time, they are best used in pairing with another storage (as an override for the default) rather than on their own.
+
+For example, we can have a feature flag which is set to disabled via a config file. This prevents users from accessing this feature. However, a developer who still wants to use this feature can send the query param as true and access it.
+
+#### 3. Growthbook
+
+GrowthBook is an open-source platform for feature flagging. It provides SDKs and a UI to manage feature flags.
+
+##### Pros
+
+1. High-level solution with UI.
+1. Supports features we'd want later like A/B testing.
+1. Can record a large number of metrics.
+1. Actively maintained and FOSS.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes.
+
+##### Cons
+
+1. Adds API calls to the server startup.
+1. Separate service that needs to be self-hosted and set up.
+1. Will need Dockerisation of the dev workflow.
+1. Might be overkill if we don't want the extra features.
+1. We might not even want to use the metrics.
+
+##### Reference
+
+- [Home](https://www.growthbook.io)
+- [Docs](https://docs.growthbook.io)
+
+#### 4. API
+
+Our API could serve as a decent source of feature flag statuses by storing them as Django model.
+
+##### Pros
+
+1. Can be managed using Django admin.
+1. No redeployment necessary, flags can be toggled remotely.
+1. Allows flags to have many sub-attributes, essentially models.
+1. The frontend depends on the API anyway, so just add more of it.
+
+##### Cons
+
+1. Need to mock the API calls in dev environments.
```

Comment:
FWIW we already do this so it doesn't seem like a big con, it's basically already done.